File: structy/matrix/2-undirected-path.py
# Write a function, undirected_path, that takes in a list of edges for an undirected graph and two nodes (node_A, node_B). The function should return a boolean indicating whether or not there exists a path between node_A and node_B.

# input: list of edges, node a, node b
# output: boolean if path exists from a to b
# convert list of edges to adj list using a for loop
# create a set to record nodes already traveled to
# create queue for depth first traversal
# while loop to iterate while queue has value
# create current with popleft
# if current == node_b return True
# for loop to iterate neighbors
# in loop check set if neighbor not included in set append and add to set
# outside both loops return False

from collections import deque

# helper to convert edges to adj_list
def build_graph(edges):
  adj_list = {}

  for (node_1, node_2) in edges:
    if node_1 not in adj_list:
      adj_list[node_1] = []

    if node_2 not in adj_list:
      adj_list[node_2] = []

    adj_list[node_1].append(node_2)
    adj_list[node_2].append(node_1)

  return adj_list

# iterative solution
def undirected_path(edges, node_A, node_B):
  graph = build_graph(edges)

  # set needed to avoid inf loop
  visited_nodes = set(node_A)
  queue = deque([node_A])
  while queue:
    current = queue.popleft()

    if current == node_B:
      return True

    for neighbor in graph[current]:
      if neighbor not in visited_nodes:
        queue.append(neighbor)
      visited_nodes.add(neighbor)

  return False


# recursive solution
def undirected_path_recur(edges, node_A, node_B):
  graph = build_graph(edges)
  return has_path(graph, node_A, node_B, set())

def has_path(graph, src, dst, visited):
  if src == dst:
    return True

  if src in visited:
    return False

  visited.add(src)

  for neighbor in graph[src]:
    if has_path(graph, neighbor, dst, visited) is True:
      return True

  return False

edges = [
  ('i', 'j'),
  ('k', 'i'),
  ('m', 'k'),
  ('k', 'l'),
  ('o', 'n')
]

print(undirected_path(edges, 'j', 'm')) # -> True
print(undirected_path_recur(edges, 'j', 'm')) # -> True

File: structy/matrix/2-undirected-path.js
// Write a function, undirectedPath, that takes in an array of edges for an undirected graph and two nodes (nodeA, nodeB). The function should return a boolean indicating whether or not there exists a path between nodeA and nodeB.

const undirectedPath = (edges, nodeA, nodeB) => {
    let graph = {}

    // populates graph using edges array
    for(let edge of edges){
        [a, b] = edge
        if(!(a in graph)){
            graph[a] = []
        }
        if (!(b in graph)){
            graph[b] = []
        }

        graph[a].push(b)
        graph[b].push(a)
    }

    // traverse using depth first algo
    let visited = new Set()
    visited.add(nodeA)
    let stack = [nodeA]

    while(stack.length > 0){
        let current = stack.pop()

        for(let neighbor of graph[current]){
            if(neighbor === nodeB){
                return true
            }else if(!visited.has(neighbor)){
                visited.add(neighbor)
                stack.push(neighbor)
            }
        }
    }

    return false
};

const edges = [
    ['i', 'j'],
    ['k', 'i'],
    ['m', 'k'],
    ['k', 'l'],
    ['o', 'n']
  ];

  console.log(undirectedPath(edges, 'j', 'm')); // -> true

  const edges1 = [
    ['i', 'j'],
    ['k', 'i'],
    ['m', 'k'],
    ['k', 'l'],
    ['o', 'n']
  ];

   console.log(undirectedPath(edges1, 'm', 'j')); // -> true
